lib/vdc/DCBOV.cpp

```diff
@@ -0,0 +1,398 @@
+    _dataVarsMap[var].SetAttribute(DC::Attribute("standard_name", DC::TEXT, var));
+    _dataVarsMap[var].SetAttribute(DC::Attribute("units", DC::TEXT, units));
+
+    return (0);
+}
+
+bool DCBOV::getDimension(string dimname, DC::Dimension &dimension) const
+{
+    map<string, DC::Dimension>::const_iterator itr;
+
+    itr = _dimsMap.find(dimname);
+    if (itr == _dimsMap.end()) return (false);
+
+    dimension = itr->second;
+    return (true);
+}
+
+std::vector<string> DCBOV::getDimensionNames() const
+{
+    map<string, DC::Dimension>::const_iterator itr;
+
+    vector<string> names;
+
+    for (itr = _dimsMap.begin(); itr != _dimsMap.end(); ++itr) { names.push_back(itr->first); }
+
+    return (names);
+}
+
+vector<string> DCBOV::getMeshNames() const
+{
+    vector<string>                         mesh_names;
+    std::map<string, Mesh>::const_iterator itr = _meshMap.begin();
+    for (; itr != _meshMap.end(); ++itr) { mesh_names.push_back(itr->first); }
+    return (mesh_names);
+}
+
+bool DCBOV::getMesh(string mesh_name, DC::Mesh &mesh) const
+{
+    map<string, Mesh>::const_iterator itr = _meshMap.find(mesh_name);
+    if (itr == _meshMap.end()) return (false);
+
+    mesh = itr->second;
+    return (true);
+}
+
+bool DCBOV::getCoordVarInfo(string varname, DC::CoordVar &cvar) const
+{
+    map<string, DC::CoordVar>::const_iterator itr;
+
+    itr = _coordVarsMap.find(varname);
+    if (itr == _coordVarsMap.end()) { return (false); }
+
+    cvar = itr->second;
+    return (true);
+}
+
+bool DCBOV::getDataVarInfo(string varname, DC::DataVar &datavar) const
+{
+    map<string, DC::DataVar>::const_iterator itr;
+
+    itr = _dataVarsMap.find(varname);
+    if (itr == _dataVarsMap.end()) { return (false); }
+
+    datavar = itr->second;
+    return (true);
+}
+
+bool DCBOV::getBaseVarInfo(string varname, DC::BaseVar &var) const
+{
+    map<string, DC::CoordVar>::const_iterator itr;
+
+    itr = _coordVarsMap.find(varname);
+    if (itr != _coordVarsMap.end()) {
+        var = itr->second;
+        return (true);
+    }
+
+    map<string, DC::DataVar>::const_iterator itr1 = _dataVarsMap.find(varname);
+    if (itr1 != _dataVarsMap.end()) {
+        var = itr1->second;
+        return (true);
+    }
+
+    return (false);
+}
+
+std::vector<string> DCBOV::getDataVarNames() const
+{
+    map<string, DC::DataVar>::const_iterator itr;
+
+    vector<string> names;
+    for (itr = _dataVarsMap.begin(); itr != _dataVarsMap.end(); ++itr) { names.push_back(itr->first); }
+    return (names);
+}
+
+std::vector<string> DCBOV::getCoordVarNames() const
+{
+    map<string, DC::CoordVar>::const_iterator itr;
+
+    vector<string> names;
+    for (itr = _coordVarsMap.begin(); itr != _coordVarsMap.end(); ++itr) { names.push_back(itr->first); }
+    return (names);
+}
+
+template<class T> bool DCBOV::_getAttTemplate(string varname, string attname, T &values) const
```

Comment:
Does the BOV format support attributes? If not why have logic for it?

Reply:
I assumed that the attributes were needed because DC has several pure-virtual functions regarding them.  (DC::getAtt, DC::getAttType, and DC::getAttNames)

Returning empty values (false, empty string, and empty vector) seems to work, so I'm inclined to call this fixed.  But I wonder if it would help other implementors in the future if these functions were not pure-virtual?

Reply:
They are needed, but if there are no attributes to return you can simply return false. There was a lot of logic here earlier that was eventually returning false. 

My thinking on virtual vs pure-virtual is that the latter forces the developer to at least think about. Clarifying the documentation could help I'm sure.